Sigma Edi (AGC 022C) (#6341)
cpinitiative/usaco-guide: solutions/gold/ac-RemainderGame.mdx

---
id: remainder-game
source: AtCoder
title: Remainder Game
author: Abhinav Singh
---

> 

Reviewer: official editorial

Author: It leads to a blank page so I didnt link it

Reviewer: you can find it [here](https://img.atcoder.jp/agc022/editorial.pdf#page=12)

(it's in the overall editorial, which is a pdf -- this is fairly common I think?)

Author: Oops didnt see the little link at thr top I saw the no editorial here in place for every question - mb will fix got lots of changes to make lol

## Explanation

Instead of tracking operations, let's first look at which **moduli** we use. Note that applying any modulus twice is identical to applying it only once, even when other moduli are used in between, so we should purchase each modulus at most once. Now, a large modulus always costs more than any combination of smaller moduli, since the worst case is purchasing every smaller modulus once:

$$2^k > \sum_{i=1}^{k-1} 2^i.$$

> This means we should prioritize optimizing the largest modulus used across all transformations $a_i \to b_i$, which motivates our greedy idea: repeatedly find the largest modulus we must buy, and buy it exactly once.

Reviewer: ```suggestion
Applying any modulus twice is identical to applying it only once, even when other moduli are used, so we should purchase each modulus at most once. Observe that a large modulus always costs more than any combination of smaller moduli, since the worst case is purchasing every smaller modulus once:

$$
2^k > \sum_{i=1}^{k-1} 2^i.
$$

This means we should prioritize optimizing the largest modulus used across all transformations $a_i \to b_i$, which motivates our greedy idea: repeatedly find the largest modulus we must buy, and buy it exactly once.
```

Hence, when processing the transformation $a_i \to b_i$, $f(a_i, b_i)$ is the minimum threshold such that the path is feasible using only operations $\leq f(a_i,b_i)$, making $K = \max_i f(a_i, b_i)$ a forced modulus in the greedy construction.

We compute these minimax distances with a Floyd–Warshall variant, replacing the standard relaxation with:

$$\texttt{dist}[i][j] = \min(\texttt{dist}[i][j],\ \max(\texttt{dist}[i][k],\ \texttt{dist}[k][j])).$$

Initially, `dist` holds the direct graph edges. We run minimax Floyd–Warshall to get all $f(u, v)$ given the current edge weights, then take $K = \max_i f(a_i, b_i)$, the largest modulus some pair still requires. If any pair is unreachable, output `-1`. If $K = 0$, every necessary modulus is already purchased, so we stop. Otherwise we add $2^K$ to the answer, reset `dist` to direct edges, and set every edge of weight $K$ to $0$, marking modulus $K$ as purchased. We repeat until done.

Each iteration permanently purchases one modulus, so there are at most $50$

> iterations.

Reviewer: this can be condensed (also you dont have to explain all the exact details of the algo, just the big picture)

## Implementation

Let $M = 50$ be the maximum value involved in the problem. Each Floyd–Warshall pass is $\mathcal{O}(M^3)$ on $M + 1$ vertices, repeated at most $M$ times, giving an
overall complexity of $\mathcal{O}(M^4)$.

> We keep `edges_by_modulus[k]`, a list of every edge with weight $k$, so that once modulus $k$ is purchased, we can mark edges using that modulus as free in $\mathcal{O}(1)$ per edge, without scanning the whole matrix.

Reviewer: we usually just outright state the time complexity (along with any terminology not defined in the statement)

if explaining the time complexity requires more work then u can put it as a note or put it in explanation

Reviewer 2: I think note is more papropriate here

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_VAL = 50;
const int NUM_NODES = MAX_VAL + 1;
const int INF = 1e9;

int dist[NUM_NODES][NUM_NODES];

> int temp_dist[NUM_NODES][NUM_NODES];

Reviewer: we normally use vectors instead of global arrays

---
id: remainder-game
source: AtCoder
title: Remainder Game
author: Abhinav Singh
---

[Official Analysis (C++)](https://img.atcoder.jp/agc022/editorial.pdf#page=12)

> ## Explanation
> 
> Applying any modulus twice is identical to applying it only once, even when other moduli are used, so we should purchase each modulus at most once. Observe that a large modulus always costs more than any combination of smaller moduli, since the worst case is purchasing every smaller modulus once:
> 
> $$
> 2^k > \sum_{i=1}^{k-1} 2^i.
> $$
> 
> This means we should prioritize optimizing the largest modulus used across all transformations $a_i \to b_i$, which motivates our greedy idea: find the largest modulus we must buy, buy it exactly once, and repeat.
> 
> To do this we build a directed graph on vertices $0, \ldots, 50$. For every value $x$ and modulus $k$, we build an edge $x \to x \bmod k$ with weight $k$. A path from $a_i$ to $b_i$ represents a sequence of operations to turn $a_i$ into $b_i$, and the moduli needed are the edge weights along the path. Define the weight of a path from $u$ to $v$, $f(u, v)$, as the minimum possible value of the maximum edge weight on such a path.
> 
> Hence, when processing a transformation $a_i \to b_i$, $f(a_i, b_i)$ is the minimum threshold such that the path is feasible using only operations $\leq f(a_i,b_i)$, so $K = \max_i f(a_i, b_i)$ must be purchased in the greedy construction.
> 
> We compute these minimax distances with a Floyd–Warshall variant, replacing the standard relaxation with:
> 
> $$\texttt{dist}[i][j] = \min(\texttt{dist}[i][j],\ \max(\texttt{dist}[i][k],\ \texttt{dist}[k][j])).$$
> 
> We run minimax Floyd–Warshall on the current graph, take $K = \max_i f(a_i, b_i)$, and repeat until $K = 0$ (or output `-1` if some pair is unreachable). Otherwise, we must purchase $K$ by addung $2^K$ to the answer, and then set every edge of weight $K$ to $0$ so future rounds treat it as free, and repeat.
> 
> Each iteration permanently purchases one modulus, so there are at most $50$ iterations.

Reviewer: ```suggestion
Applying any modulus twice is identical to applying it only once, even when other moduli are used, so we should purchase each modulus at most once. Observe that a large modulus always costs more than any combination of smaller moduli, since the worst case is purchasing every smaller modulus once:

$$
2^k > \sum_{i=1}^{k-1} 2^i.
$$

This means we should prioritize optimizing the largest modulus used across all transformations $a_i \to b_i$, which motivates our greedy idea: find the largest modulus we must buy, buy it exactly once, and repeat.

To do this, we build a directed graph on vertices $0, \ldots, 50$. For every value $x$ and modulus $k$, we build an edge $x \to x \bmod k$ with weight $k$. A path from $a_i$ to $b_i$ represents a sequence of operations to turn $a_i$ into $b_i$, and the moduli needed are the edge weights along the path. 

Define the weight of a path from $u$ to $v$, $f(u, v)$, as the minimum possible value of the maximum edge weight on such a path. Hence, when processing a transformation $a_i \to b_i$, $f(a_i, b_i)$ is the minimum threshold such that the path is feasible using only operations $\leq f(a_i,b_i)$, so $K = \max_i f(a_i, b_i)$ must be purchased in the greedy construction. Note that $f(u, v) = 0$ when $a_i$ can reach $b_i$ in the graph.

We compute these distances with a Floyd-Warshall variant, replacing the standard relaxation with:

$$
\texttt{dist}[i][j] = \min(\texttt{dist}[i][j],\ \max(\texttt{dist}[i][k],\ \texttt{dist}[k][j])).
$$

As described earlier, it is optimal to purchase the largest necessary modulus. To do so, we calculate $K = \max_i f(a_i, b_i)$. If it is impossible to transform any pair of $a_i$ and $b_i$, we print `-1`. Otherwise, we purchase $K$, and set every edge with weight $K$ to $0$. Note that this requires recalculating $f(a_i, b_i)$ later.

Each iteration permanently purchases one modulus, so there are at most $50$ iterations.
```

at least this is how i'd rewrite some parts -- it's mostly good as is

(in particular, I removed the 'minimax' term, since that has a pretty specific meaning I think)

## Implementation

> **Time Complexity**: Let $M = 50$ be the maximum value involved in the problem. Each Floyd–Warshall pass is $\mathcal{O}(M^3)$ on $M + 1$ vertices, and it is repeated at most $M$ times, so the overall complexity it $\mathcal{O}(M^4)$.

Reviewer: ```suggestion
**Time Complexity**: $\mathcal{O}(M^4)$, where $M=\max_i(a_i)$.
```

i don't think you need to explain this in depth, since it's just 3 for loops and u explain # of iterations

<Note>

We keep `edges_by_modulus[k]`, a list of every edge with weight $k$, so that once modulus $k$ is purchased, we can mark edges using that modulus as free in
$\mathcal{O}(1)$ per edge, without scanning the whole matrix.

</Note>

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX_VAL = 50;
const int NUM_NODES = MAX_VAL + 1;
const int INF = 1e9;

int dist[NUM_NODES][NUM_NODES];

> int temp_dist[NUM_NODES][NUM_NODES];

Reviewer: use vectors instead ?

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;

	vector<int> a(n), b(n);
	for (int &x : a) cin >> x;
	for (int &x : b) cin >> x;

> 	vector<pair<int, int>> edges_by_modulus[NUM_NODES];

Reviewer: should turn this into a 2d vector (VLAs are generally bad practice, esp cuz some OJs enforce stack memory limits)

Author: sorry whats a VLA

Reviewer: oh nvm i thought this was a variable length array but NUM_NODES is fixed

regardless usually we'd use a 2d vector

---
id: ac-RemainderGame
source: AtCoder Grand Contest 022
title: Remainder Game
author: Abhinav Singh
---

## Explanation

> Since modulo can never increase a value, if $a_i < b_i$ for some $i$, the answer is immediately $-1$.

Reviewer 2: This is unneccessary no?  the described algorthm will simply reply $-1$ in this case

Author: oop

> Instead of tracking operations, let's track which **moduli** we use. Using modulus $k$ costs $2^k$, and lets us reduce any element mod $k$

Reviewer 2: We still need to constructively track operations when we calculate answer.

So get rid of this whole sentence, since it just restates the problem.

If you wanna tell me what intuition you're trying to convey here lmk and we can sort out a more concise way

Author: yeah mb i shd say lets look at which moduli instead of lets track thats on me

As $$2^k > \sum_{i=1}^{k-1} 2^i,$$

> a single larger modulus always costs more than every smaller modulus combined. This means that for each transformation from $a_i \to b_i$, what matters isn't the total number of moduli used, but rather the largest one. We want the path that minimizes the maximum edge weight from $a_i$ to $b_i$ for each $i$ (Note that applying $k$ twice is identical to applying it once: $$(x \bmod k) \bmod k = x \bmod k,$$ so an optimal solution never needs to purchase the same modulus twice). This motivates a greedy idea: repeatedly find the largest modulus every solution is forced to buy.

Reviewer 2: Move the clause "Note that applying $k$ twice is identical to applying it once: $(x \bmod k) \bmod k = x \bmod k,$ so an optimal solution never needs to purchase the same modulus twice" before this sentence.

Now, you can say "A large modulus always costs more than any combination of smaller modulus's", as [your equation on this line]

that way it flows smoother and eliminates the doubts of if you're gonna buy a lot of copies of a small modulus and exceed cost of big one

I'd do something like this but the standard disclaimer that i admit norz follows so i encourage rewriting

```suggestion
Applying any modulus twice is identical to applying it only once, so we should purchase each modulus at most once.
Now, note that a large modulus always costs more than any combination of smaller moduli, as the worst combination is to purchase each moduli only once, and we have:

$$2^k > \sum_{i=1}^{k-1} 2^i.$$

This means that we should prioritize optimizing the largest modulus used in all transformations $a_i \to b_i$ each time, which motivates a greedy idea:
repeatedly finding the largest modulus every solution is forced to buy, and buying that modulus exactly once.
```

Let's build a directed graph on vertices $0, \ldots, 50$. For every value $x$ and modulus $k$, add an edge

$$x \to x \bmod k$$

with weight $k$. A path from $a_i$ to $b_i$ represents a sequence of operations which turn $a_i$ into $b_i$, and the moduli needed are the edge weights on that path. Define a *path's* weight as its **maximum** edge weight, and let

> $$f(u, v) = \min_{\text{paths } u \to v} \ \max(\text{edge weight}).$$

Reviewer 2: 1: inline the first display equation. Also could explicitly state what the graph is supposed to represent.
2. No need to lay out this last f equation, just say "define a path from $i$ to $j$'s weight, $f(i, j)$,..." in previous paragraph

> Hence, when processing moduli for the path $a_i \to b_i$, $f(a_i, b_i) = K$ means $K$ is the minimum threshold such that the path becomes feasible using only operations $\le K$, so in the across all the paths this is the first point where this pair cannot be satisfied without introducing modulo $K$, making it a forced contribution in the greedy construction.

Reviewer 2: ```suggestion
Hence, when processing moduli for the path $a_i \to b_i$, $k = f(a_i, b_i)$ is the minimum threshold such that the path becomes feasible using only operations $\leq f(a_i, b_i)$, making the maximum $k$ across all paths $a_i \to b_i$ a forced contribution in the greedy construction.
```

We can compute these minimax distances with a Floyd–Warshall variant:

> ```cpp

Reviewer 2: Either languagesection or convert to LaTeX

dist[i][j] = min(dist[i][j], max(dist[i][k], dist[k][j]));
```

Initially `dist` holds the direct graph edges, and we repeat the following until done. We run minimax Floyd–Warshall to get all $f(u, v)$ given the current edge

> weights, then take $K = \max_i f(a_i, b_i)$, which is the largest modulus some pair still requires. If any pair is unreachable, output `-1`. If $K = 0$, every necessary modulus is already purchased, so we stop. Otherwise we add $2^K$ to the answer and set every edge of weight $K$ to $0$, marking modulus $K$ as purchased and free for all future paths.

Reviewer 2: ```suggestion
Initially, `dist` holds direct graph edges. We run minimax Floyd–Warshall to get all $f(u, v)$ given the current edge
weights, then take $K = \max_i f(a_i, b_i)$, which is the largest modulus some pair still requires. If any pair is unreachable, output `-1`. If $K = 0$, every necessary modulus is already purchased, so we stop. Otherwise we add $2^K$ to the answer, reset `dist` to direct edges, and set every edge of weight $K$ to $0$, marking modulus $K$ as purchased and free for all future paths.
We repeat the above until done, i.e. the maximum required value is $0$.
```

Each iteration permanently uses one modulus, so there are at most $50$ iterations.

## Implementation

Each Floyd–Warshall pass is $\mathcal{O}(51^3)$ on $51$ vertices, repeated at most
$50$ times, giving an overall complexity of

> $$\mathcal{O}(50 \cdot 51^3) = \mathcal{O}(50^4).$$

Reviewer 2: You can't quote constants in big-O notation as it collaspses to O(1)

so like add a sentence like "Let $\mathtt{maxn}$ be the maximum number used in the problem" and use that.

We keep `edges_by_modulus[k]`, a list of every edge with weight $k$, so that once
modulus $k$ is purchased we can mark exactly those edges as free in $\mathcal{O}(1)$
per edge, without scanning the whole adjacency matrix.

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

> const int num_values = 51;

Reviewer 2: Add a separate constant `MAX_VALUES=50` and use `<=` for that.  That way it's clearer when you enumerate over `MAX_VALUES` you're enumerating actual values not indices.

> const int inf = 1e9;

Reviewer 2: Capitalize your constants

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int n;
	cin >> n;

	vector<int> a(n), b(n);

> 	for (int &x : a) cin >> x;

Reviewer 2: Orz use of foreach iterator

	for (int &x : b) cin >> x;

> 	int dist[num_values][num_values];

Reviewer 2: Isn't it preferrable to declare arrays globally?  Woud like another perspective tho

	vector<pair<int, int>> edges_by_modulus[num_values];

	for (int i = 0; i < num_values; i++) {

> 		for (int j = 0; j < num_values; j++) dist[i][j] = inf;

Reviewer 2: Put braces around the substatement

<img width="535" height="308" alt="Image" src="https://github.com/user-attachments/assets/ce4d9dd3-f112-4d8c-aa8e-f31f37f8d12f" />

		dist[i][i] = 0;
	}

	// Add an edge x -> x % k with weight k, and store it for fast lookup later
	for (int x = 0; x < num_values; x++) {

> 		for (int k = 1; k <= 50; k++) {

Reviewer 2: Hardcoding 50... use constant

			int y = x % k;
			dist[x][y] = min(dist[x][y], k);
			edges_by_modulus[k].push_back({x, y});
		}
	}

	long long ans = 0;

	while (true) {

> 		int temp_dist[num_values][num_values];

Reviewer 2: Declare globally, since memcpy would overwrite anyways
